client/admin/chat-channels.tsx

```diff
@@ -0,0 +1,269 @@
+import {
+  areUserEntriesEqual,
+  sortUserEntries,
+  useUserEntriesSelector,
+} from '../users/sorted-user-ids'
+
+const SEARCH_CHANNELS_LIMIT = 10
+const CHANNEL_MESSAGES_LIMIT = 50
+
+const Container = styled.div`
+  max-width: 960px;
+  height: 100%;
+  padding: 0 16px;
+
+  display: flex;
+  flex-direction: column;
+`
+
+const PageHeadline = styled.div`
+  ${headline5};
+  margin-top: 16px;
+  margin-bottom: 8px;
+`
+
+const StyledSearchInput = styled(SearchInput)`
+  width: 256px;
+`
+
+const SearchContainer = styled.div`
+  height: 88px;
+  margin: 16px 0;
+`
+
+const NoResults = styled.div`
+  ${subtitle1};
+
+  color: ${colorTextFaint};
+`
+
+const ErrorText = styled.div`
+  ${subtitle1};
+
+  color: ${colorError};
+`
+
+const StyledCarousel = styled(Carousel)`
+  max-width: 960px;
+`
+
+const ChannelThumbnail = styled(Card)`
+  width: 420px;
+  padding: 16px;
+
+  flex-shrink: 0;
+  display: flex;
+  flex-direction: row;
+  align-items: center;
+
+  &:not(:first-child) {
+    margin-left: 16px;
+  }
+`
+
+const StyledChannelIcon = styled(ChannelIcon)`
+  width: 56px;
+  height: auto;
+  flex-shrink: 0;
+`
+
+const ChannelInfoContainer = styled.div`
+  flex-grow: 1;
+  display: flex;
+  flex-direction: column;
+  margin: 0 8px;
+  overflow: hidden;
+`
+
+const ChannelInfoName = styled(Headline6)`
+  ${singleLine};
+`
+
+const ChannelHeadline = styled.div`
+  ${headline6};
+  margin-bottom: 8px;
+`
+
+const ChannelContainer = styled.div`
+  width: 100%;
+  overflow: hidden;
+
+  flex-grow: 1;
+  display: flex;
+`
+
+const StyledMessageList = styled(MessageList)`
+  flex-grow: 1;
+  max-width: 960px;
+  min-width: 320px;
+`
+
+export function AdminChatChannels() {
+  const dispatch = useAppDispatch()
+  const [channels, setChannels] = useState<ChannelInfo[]>()
+  const [currentPage, setCurrentPage] = useState(-1)
+  const [hasMoreChannels, setHasMoreChannels] = useState(true)
+
+  const [isLoadingMoreChannels, setIsLoadingMoreChannels] = useState(false)
+  const [searchError, setSearchError] = useState<Error>()
+  const [searchQuery, setSearchQuery] = useLocationSearchParam('q')
+
+  const carouselRef = useRef(null)
+  const debouncedSearchRef = useRef(
+    debounce((searchQuery: string) => {
+      // Just need to clear the search results here and let the carousel initiate the network
+      // request.
+      setSearchQuery(searchQuery)
+      setSearchError(undefined)
+      setChannels(undefined)
+      setCurrentPage(-1)
+      setHasMoreChannels(true)
+      if (carouselRef.current) {
+        // TODO(2Pac): Fix the types once the carousel is ported to TypeScript.
+        ;(carouselRef.current as any).reset()
+      }
+    }, 100),
+  )
+
+  const onSearchChange = useStableCallback((searchQuery: string) => {
+    debouncedSearchRef.current(searchQuery)
+  })
+
+  const onLoadMoreChannels = useStableCallback(() => {
+    setIsLoadingMoreChannels(true)
+    setCurrentPage(currentPage + 1)
+
+    fetchJson<ChannelInfo[]>(
```

Comment:
You need some way to abort this, as-is you can screw up the results requesting the next page and then triggering a new search probably